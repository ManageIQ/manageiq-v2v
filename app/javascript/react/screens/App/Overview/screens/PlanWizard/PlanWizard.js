import React from 'react';
import PropTypes from 'prop-types';
import { bindMethods, noop, Button, Icon, Wizard } from 'patternfly-react';
import { createMigrationPlans } from './helpers';
import PlanWizardBody from './PlanWizardBody';

const planWizardSteps = [
  'planWizardGeneralStep',
  'planWizardVMStep',
  'planWizardOptionsStep',
  'planWizardResultsStep'
];

class PlanWizard extends React.Component {
  constructor() {
    super();
    this.state = { activeStepIndex: 0 };
    bindMethods(this, ['prevStep', 'nextStep', 'goToStep']);
  }

  prevStep() {
    const { resetVmStepAction } = this.props;
    const { activeStepIndex } = this.state;

    if (activeStepIndex === 1) {
      // reset all vm step values if going back from that step
      resetVmStepAction();
    }
    this.setState({ activeStepIndex: Math.max(activeStepIndex - 1, 0) });
  }

  nextStep() {
    const { activeStepIndex } = this.state;
    const {
      planWizardGeneralStep,
      planWizardVMStep,
      planWizardOptionsStep,
      setPlansBodyAction,
      setPlanScheduleAction,
      setMigrationsFilterAction
    } = this.props;

    if (activeStepIndex === 2) {
      const plansBody = createMigrationPlans(
        planWizardGeneralStep,
        planWizardVMStep
      );

      setPlanScheduleAction(
        planWizardOptionsStep.values.migration_plan_choice_radio
      );
      setPlansBodyAction(plansBody);

      if (
        planWizardOptionsStep.values.migration_plan_choice_radio ===
        'migration_plan_now'
      ) {
        setMigrationsFilterAction('Migration Plans in Progress');
      } else if (
        planWizardOptionsStep.values.migration_plan_choice_radio ===
        'migration_plan_later'
      ) {
        setMigrationsFilterAction('Migration Plans Not Started');
      }
    }

    this.setState({
      activeStepIndex: Math.min(activeStepIndex + 1, planWizardSteps.length - 1)
    });
  }

  goToStep(activeStepIndex) {
    this.setState({ activeStepIndex });
  }

  render() {
    const {
      hidePlanWizard,
      hidePlanWizardAction,
      planWizardExitedAction,
      planWizardGeneralStep,
      planWizardVMStep,
      planWizardOptionsStep
    } = this.props;

    const { activeStepIndex, plansBody } = this.state;
    const activeStep = (activeStepIndex + 1).toString();
    const onFirstStep = activeStepIndex === 0;
    const onFinalStep = activeStepIndex === planWizardSteps.length - 1;

    const currentStepProp = !onFinalStep && planWizardSteps[activeStepIndex];
    const currentStepForm = !onFinalStep && this.props[currentStepProp];

    const disableNextStep =
      (!onFinalStep && !!currentStepForm.syncErrors) ||
      (activeStepIndex === 1 &&
        (!this.props.planWizardVMStep.values ||
          !this.props.planWizardVMStep.values.selectedVms ||
          this.props.planWizardVMStep.values.selectedVms.length === 0));

    return (
      <Wizard
        show={!hidePlanWizard}
        onClose={hidePlanWizardAction}
        onExited={planWizardExitedAction}
      >
        <Wizard.Header
          onClose={hidePlanWizardAction}
          title={__('Migration Plan Wizard')}
        />

        <Wizard.Body>
          <PlanWizardBody
            loaded
            activeStepIndex={activeStepIndex}
            activeStep={activeStep}
            goToStep={this.goToStep}
            disableNextStep={disableNextStep}
            plansBody={plansBody}
            planWizardGeneralStep={planWizardGeneralStep}
            planWizardVMStep={planWizardVMStep}
            planWizardOptionsStep={planWizardOptionsStep}
          />
        </Wizard.Body>

        <Wizard.Footer className="wizard-pf-footer">
          <Button
            bsStyle="default"
            className="btn-cancel"
            onClick={hidePlanWizardAction}
          >
            {__('Cancel')}
          </Button>
          <Button
            bsStyle="default"
            onClick={this.prevStep}
            disabled={onFirstStep || onFinalStep}
          >
            <Icon type="fa" name="angle-left" />
            {__('Back')}
          </Button>
          <Button
            bsStyle="primary"
            onClick={onFinalStep ? hidePlanWizardAction : this.nextStep}
            disabled={disableNextStep}
          >
            {onFinalStep
              ? __('Close')
              : activeStepIndex === 2
                ? __('Create')
                : __('Next')}
            <Icon type="fa" name="angle-right" />
          </Button>
        </Wizard.Footer>
      </Wizard>
    );
  }
}
PlanWizard.propTypes = {
  hidePlanWizard: PropTypes.bool,
  hidePlanWizardAction: PropTypes.func,
  planWizardExitedAction: PropTypes.func,
  planWizardGeneralStep: PropTypes.object,
  planWizardVMStep: PropTypes.object,
  planWizardOptionsStep: PropTypes.object,
  setPlansBodyAction: PropTypes.func,
  setPlanScheduleAction: PropTypes.func,
  resetVmStepAction: PropTypes.func,
  setMigrationsFilterAction: PropTypes.func
};
PlanWizard.defaultProps = {
  hidePlanWizard: true,
  hidePlanWizardAction: noop,
  planWizardExitedAction: noop,
  planWizardGeneralStep: {},
  planWizardVMStep: {},
  planWizardOptionsStep: {},
  setPlansBodyAction: noop,
  setPlanScheduleAction: noop,
  resetVmStepAction: noop
};
export default PlanWizard;
