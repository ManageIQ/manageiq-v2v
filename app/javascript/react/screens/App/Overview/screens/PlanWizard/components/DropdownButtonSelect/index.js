import DropdownButtonSelect from './DropdownButtonSelect';

export default DropdownButtonSelect;
