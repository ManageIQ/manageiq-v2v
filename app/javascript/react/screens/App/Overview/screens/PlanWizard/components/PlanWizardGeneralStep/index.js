import { connect } from 'react-redux';
import PlanWizardGeneralStep from './PlanWizardGeneralStep';

const mapStateToProps = ({ overview }) => ({
  transformationMappings: overview.transformationMappings,
  initialValues: { infrastructure_mapping: overview.planWizardId }
});

export default connect(mapStateToProps)(PlanWizardGeneralStep);
