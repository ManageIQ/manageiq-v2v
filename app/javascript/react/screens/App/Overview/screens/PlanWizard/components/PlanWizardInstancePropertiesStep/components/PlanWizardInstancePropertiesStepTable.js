import React from 'react';
import PropTypes from 'prop-types';
import orderBy from 'lodash.orderby';
import * as sort from 'sortabular';
import * as resolve from 'table-resolver';
import { compose } from 'recompose';
import { paginate, Grid, PaginationRow, Table, PAGINATION_VIEW, Icon, Button, FormControl } from 'patternfly-react';

// Temporary import while https://github.com/patternfly/patternfly-react/issues/535 is open:
import TableInlineEditRow from './TableInlineEditRow/TableInlineEditRow';

class PlanWizardInstancePropertiesStepTable extends React.Component {
  state = {
    // Sort the first column in an ascending way by default.
    sortingColumns: {
      name: {
        direction: Table.TABLE_SORT_DIRECTION.ASC,
        position: 0
      }
    },

    editing: false,

    // pagination default states
    pagination: {
      page: 1,
      perPage: 5,
      perPageOptions: [5, 10, 15]
    },

    // page input value
    pageChangeValue: 1
  };

  componentWillUnmount() {
    const { input, rows } = this.props;
    const minimalInstancePropertiesRows = [];

    rows.forEach(row => {
      const { id, osp_security_group, osp_flavor } = row;
      minimalInstancePropertiesRows.push({
        vm_id: id,
        osp_security_group_id: osp_security_group.id,
        osp_flavor_id: osp_flavor.id
      });
    });

    input.onChange({ rows: minimalInstancePropertiesRows, updatedInstancePropertiesRowOnStandby: {} });
  }

  // enables our custom header formatters extensions to reactabular
  customHeaderFormatters = Table.customHeaderFormattersDefinition;
  inlineEditController = () => {
    const { rows, instancePropertiesRowsAction, input } = this.props;
    return {
      isEditing: ({ rowData }) => rowData.id === input.value.updatedInstancePropertiesRowOnStandby.id,
      onActivate: ({ rowData }) => {
        this.setState({ editing: true });
        input.onChange({ updatedInstancePropertiesRowOnStandby: rowData });
      },
      onConfirm: ({ rowData }) => {
        this.setState({ editing: false });

        const updatedRows = rows.map(
          row => (row.id === rowData.id ? input.value.updatedInstancePropertiesRowOnStandby : row)
        );
        instancePropertiesRowsAction(updatedRows);
        input.onChange({ updatedInstancePropertiesRowOnStandby: {} });
      },
      onCancel: ({ rowData }) => {
        this.setState({ editing: false });
        input.onChange({ updatedInstancePropertiesRowOnStandby: {} });
      },
      onChange: (e, { rowData, property }) => {
        const updatedInstanceProp = {
          ...rowData[property],
          name: e.target.options[e.target.selectedIndex].text,
          id: e.target.value
        };

        const updatedRowdata = {
          ...input.value.updatedInstancePropertiesRowOnStandby,
          [property]: updatedInstanceProp
        };

        input.onChange({ updatedInstancePropertiesRowOnStandby: updatedRowdata });
      }
    };
  };

  inlineEditFormatter = Table.inlineEditFormatterFactory({
    isEditing: additionalData => this.inlineEditController().isEditing(additionalData),
    renderValue: (value, additionalData) => (
      <td className="editable">
        <span className="static">
          {additionalData.property === 'osp_security_group'
            ? additionalData.rowData.osp_security_group.name
            : additionalData.rowData.osp_flavor.name}
        </span>
      </td>
    ),
    renderEdit: (value, additionalData) => {
      const { tenantsWithAttributesById, destinationTenantIdsBySourceClusterId, input } = this.props;
      const { optionsAttribute } = additionalData.column.cell.inlineEditSelect;
      const clusterId = additionalData.rowData.ems_cluster_id;
      const tenantId = destinationTenantIdsBySourceClusterId[clusterId];
      const tenant = tenantId && tenantsWithAttributesById[tenantId];
      const allFitForVM = () => {
        const { bestFitFlavors } = this.props;
        const flavorsSlugPrefix = 'flavors/';
        const allFitIds = bestFitFlavors
          .find(flavor => flavor.source_href_slug === `vms/${additionalData.rowData.id}`)
          .all_fit.map(flavorSlug => flavorSlug.slice(flavorsSlugPrefix.length));
        return tenant.flavors.filter(flavor => allFitIds.indexOf(flavor.id) > -1);
      };
      const options = tenant ? (optionsAttribute === 'flavors' ? allFitForVM() : tenant[optionsAttribute]) : [];
      return (
        <td className="editable editing">
          <FormControl
            componentClass="select"
            defaultValue={
              additionalData.property === 'osp_security_group'
                ? input.value.updatedInstancePropertiesRowOnStandby.osp_security_group.id
                : input.value.updatedInstancePropertiesRowOnStandby.osp_flavor.id
            }
            onChange={e => this.inlineEditController().onChange(e, additionalData)}
          >
            {options.map(opt => (
              <option value={opt.id} name={opt.name} key={opt.id}>
                {opt.name}
              </option>
            ))}
          </FormControl>
        </td>
      );
    }
  });

  inlineEditButtonsFormatter = Table.inlineEditFormatterFactory({
    isEditing: additionalData => this.state.editing,
    renderValue: (value, additionalData) => (
      <td style={{ padding: '2px' }}>
        <Button bsStyle="default" onClick={() => this.inlineEditController().onActivate(additionalData)}>
          <Icon type="pf" name="edit" />
        </Button>
      </td>
    ),
    renderEdit: (value, additionalData) => (
      <td style={{ padding: '2px' }}>
        <Button bsStyle="default" disabled>
          <Icon type="pf" name="edit" />
        </Button>
      </td>
    )
  });

  getColumns = () => {
    const getSortingColumns = () => this.state.sortingColumns || {};

    const sortableTransform = sort.sort({
      getSortingColumns,
      onSort: selectedColumn => {
        this.setState({
          sortingColumns: sort.byColumn({
            sortingColumns: this.state.sortingColumns,
            sortingOrder: Table.defaultSortingOrder,
            selectedColumn
          })
        });
      },
      // Use property or index dependening on the sortingColumns structure specified
      strategy: sort.strategies.byProperty
    });

    const sortingFormatter = sort.header({
      sortableTransform,
      getSortingColumns,
      strategy: sort.strategies.byProperty
    });

    return [
      {
        property: 'name',
        header: {
          label: __('VM Name'),
          props: {
            index: 0,
            rowSpan: 1,
            colSpan: 1
          },
          transforms: [sortableTransform],
          formatters: [sortingFormatter],
          customFormatters: [Table.sortableHeaderCellFormatter]
        },
        cell: {
          props: {
            index: 0
          },
          formatters: [Table.tableCellFormatter]
        }
      },
      {
        property: 'cluster',
        header: {
          label: __('Source Cluster'),
          props: {
            index: 1,
            rowSpan: 1,
            colSpan: 1
          },
          transforms: [sortableTransform],
          formatters: [sortingFormatter],
          customFormatters: [Table.sortableHeaderCellFormatter]
        },
        cell: {
          props: {
            index: 1
          },
          formatters: [Table.tableCellFormatter]
        }
      },
      {
        property: 'target_cluster_name',
        header: {
          label: __('Target Provider'),
          props: {
            index: 2,
            rowSpan: 1,
            colSpan: 1
          },
          transforms: [sortableTransform],
          formatters: [sortingFormatter],
          customFormatters: [Table.sortableHeaderCellFormatter]
        },
        cell: {
          props: {
            index: 2
          },
          formatters: [Table.tableCellFormatter]
        }
      },
      {
        property: 'allocated_size',
        header: {
          label: __('Allocated Size'),
          props: {
            index: 3,
            rowSpan: 1,
            colSpan: 1
          },
          transforms: [sortableTransform],
          formatters: [sortingFormatter],
          customFormatters: [Table.sortableHeaderCellFormatter]
        },
        cell: {
          props: {
            index: 3
          },
          formatters: [Table.tableCellFormatter]
        }
      },
      {
        property: 'osp_security_group',
        header: {
          label: __('OpenStack Security Group'),
          props: {
            index: 4,
            rowSpan: 1,
            colSpan: 1
          },
          transforms: [sortableTransform],
          formatters: [sortingFormatter],
          customFormatters: [Table.sortableHeaderCellFormatter]
        },
        cell: {
          props: {
            index: 4
          },
          formatters: [this.inlineEditFormatter],
          inlineEditSelect: {
            // Custom property for inlineEditFormatter
            optionsAttribute: 'security_groups'
          }
        }
      },
      {
        property: 'osp_flavor',
        header: {
          label: __('OpenStack Flavor'),
          props: {
            index: 5,
            rowSpan: 1,
            colSpan: 1
          },
          transforms: [sortableTransform],
          formatters: [sortingFormatter],
          customFormatters: [Table.sortableHeaderCellFormatter]
        },
        cell: {
          props: {
            index: 5
          },
          formatters: [this.inlineEditFormatter],
          inlineEditSelect: {
            // Custom property for inlineEditFormatter
            optionsAttribute: 'flavors'
          }
        }
      },
      {
        property: 'actions',
        header: {
          label: '',
          props: {
            index: 6
          },
          formatters: [Table.actionHeaderCellFormatter]
        },
        cell: {
          props: {
            index: 6
          },
          formatters: [this.inlineEditButtonsFormatter]
        }
      }
    ];
  };

  onFirstPage = () => {
    this.setPage(1);
  };
  onLastPage = () => {
    const { page } = this.state.pagination;
    const totalPages = this.totalPages();
    if (page < totalPages) {
      this.setPage(totalPages);
    }
  };
  onNextPage = () => {
    const { page } = this.state.pagination;
    if (page < this.totalPages()) {
      this.setPage(this.state.pagination.page + 1);
    }
  };
  onPageInput = e => {
    this.setState({ pageChangeValue: e.target.value });
  };
  onPerPageSelect = (eventKey, e) => {
    const newPaginationState = Object.assign({}, this.state.pagination);
    newPaginationState.perPage = eventKey;
    newPaginationState.page = 1;
    this.setState({ pagination: newPaginationState });
  };
  onPreviousPage = () => {
    if (this.state.pagination.page > 1) {
      this.setPage(this.state.pagination.page - 1);
    }
  };

  onSubmit = () => {
    this.setPage(this.state.pageChangeValue);
  };

  setPage = value => {
    const page = Number(value);
    if (!Number.isNaN(value) && value !== '' && page > 0 && page <= this.totalPages()) {
      const newPaginationState = Object.assign({}, this.state.pagination);
      newPaginationState.page = page;
      this.setState({ pagination: newPaginationState, pageChangeValue: page });
    }
  };

  currentRows = () => {
    const { sortingColumns, pagination } = this.state;
    const { rows } = this.props;

    return compose(
      paginate(pagination),
      sort.sorter({
        columns: this.getColumns(),
        sortingColumns,
        sort: orderBy,
        strategy: sort.strategies.byProperty
      })
    )(rows);
  };

  totalPages = () => {
    const { rows } = this.props;
    const { perPage } = this.state.pagination;
    return Math.ceil(rows.length / perPage);
  };

  render() {
    const { pagination, sortingColumns, pageChangeValue } = this.state;

    const sortedPaginatedRows = this.currentRows();

    return (
      <Grid fluid>
        <h2>{__('VMs Selected for Migration')}</h2>
        <br />
        <Table.PfProvider
          striped
          bordered
          hover
          dataTable
          inlineEdit
          columns={this.getColumns()}
          components={{
            header: {
              cell: cellProps =>
                this.customHeaderFormatters({
                  cellProps,
                  columns: this.getColumns(),
                  sortingColumns,
                  rows: sortedPaginatedRows.rows
                })
            },
            body: {
              // TODO we should use Table.InlineEditRow when patternfly-react#535 is fixed
              row: TableInlineEditRow
            }
          }}
        >
          <Table.Header headerRows={resolve.headerRows({ columns: this.getColumns() })} />
          <Table.Body
            rows={sortedPaginatedRows.rows || []}
            rowKey="id"
            onRow={(rowData, { rowIndex }) => ({
              role: 'row',
              isEditing: () => this.inlineEditController().isEditing({ rowData }),
              onCancel: () => this.inlineEditController().onCancel({ rowData, rowIndex }),
              onConfirm: () => this.inlineEditController().onConfirm({ rowData, rowIndex }),
              last: rowIndex === sortedPaginatedRows.length - 1,
              tableRendersInModal: true
            })}
          />
        </Table.PfProvider>
        <PaginationRow
          viewType={PAGINATION_VIEW.TABLE}
          pagination={pagination}
          pageInputValue={pageChangeValue}
          amountOfPages={sortedPaginatedRows.amountOfPages}
          itemCount={sortedPaginatedRows.itemCount}
          itemsStart={sortedPaginatedRows.itemsStart}
          itemsEnd={sortedPaginatedRows.itemsEnd}
          onPerPageSelect={this.onPerPageSelect}
          onFirstPage={this.onFirstPage}
          onPreviousPage={this.onPreviousPage}
          onPageInput={this.onPageInput}
          onNextPage={this.onNextPage}
          onLastPage={this.onLastPage}
          onSubmit={this.onSubmit}
        />
      </Grid>
    );
  }
}
PlanWizardInstancePropertiesStepTable.propTypes = {
  rows: PropTypes.array,
  tenantsWithAttributesById: PropTypes.object,
  destinationTenantIdsBySourceClusterId: PropTypes.object,
  instancePropertiesRowsAction: PropTypes.func,
  input: PropTypes.object,
  bestFitFlavors: PropTypes.array
};
PlanWizardInstancePropertiesStepTable.defaultProps = {
  rows: [],
  tenantsWithAttributesById: {},
  destinationTenantIdsBySourceClusterId: {}
};
export default PlanWizardInstancePropertiesStepTable;
