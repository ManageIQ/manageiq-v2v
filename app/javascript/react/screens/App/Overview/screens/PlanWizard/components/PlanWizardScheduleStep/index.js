import { connect } from 'react-redux';
import PlanWizardScheduleStep from './PlanWizardScheduleStep';
import { getCurrentTargetProvider } from '../../PlanWizardSelectors';

const mapStateToProps = ({ overview: { transformationMappings }, form }) => ({
  targetProvider: getCurrentTargetProvider(form, transformationMappings),
  migration_plan_choice_radio:
    form.planWizardScheduleStep &&
    form.planWizardScheduleStep.values &&
    form.planWizardScheduleStep.values.migration_plan_choice_radio,
  migration_plan_type_radio:
    form.planWizardScheduleStep &&
    form.planWizardScheduleStep.values &&
    form.planWizardScheduleStep.values.migration_plan_type_radio,
  initialValues: {
    migration_plan_choice_radio: 'migration_plan_later',
    migration_plan_type_radio: 'migration_type_cold'
  }
});

export default connect(mapStateToProps)(PlanWizardScheduleStep);
