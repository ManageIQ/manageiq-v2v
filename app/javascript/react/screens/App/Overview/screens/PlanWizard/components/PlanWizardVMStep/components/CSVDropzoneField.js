import React from 'react';
import PropTypes from 'prop-types';
import csv from 'csv';
import { bindMethods, Button, EmptyState } from 'patternfly-react';
import Dropzone from 'react-dropzone';

// Unfortunately, this is the recommended way to trigger the file dialog programmatically.
// https://github.com/react-dropzone/react-dropzone/tree/master/examples/File%20Dialog
let dropzoneRef;

class CSVDropzoneField extends React.Component {
  constructor() {
    super();
    bindMethods(this, [
      'onFileDrop',
      'handleUnparsedFile',
      'trimWhiteSpaces',
      'mapCSVColumnNameToKey'
    ]);
  }

  onFileDrop(acceptedFiles, rejectedFiles) {
    if (acceptedFiles && acceptedFiles.length > 0) {
      this.handleUnparsedFile(acceptedFiles[0]);
    } else {
      // eslint-disable-next-line no-restricted-globals
      alert('Invalid file extension. Only .csv files are accepted.');
    }
  }

  handleUnparsedFile(fileHandle) {
    const { onCSVParseSuccess, onCSVParseFailure } = this.props;
    if (fileHandle) {
      const reader = new FileReader();
      reader.onload = () => {
        csv.parse(reader.result, (err, csvRows) => {
          if (csvRows && csvRows.length > 0) {
            csvRows = this.trimWhiteSpaces(csvRows);
            const headerRow = this.mapCSVColumnNameToKey(csvRows[0]);
            if (!headerRow.find(element => element === 'name')) {
              onCSVParseFailure(
                __(
                  "Error: Required column 'Name' does not exist in the .CSV file"
                )
              );
              return;
            }
            const rowObjects = csvRows.map(row =>
              headerRow.reduce(
                (rowObject, key, index) => ({
                  ...rowObject,
                  [key]: row[index]
                }),
                {}
              )
            );
            onCSVParseSuccess(rowObjects);
          } else {
            onCSVParseFailure(
              err ? `${err}` : __('Error: Possibly a blank .CSV file')
            );
          }
        });
      };
      reader.readAsBinaryString(fileHandle);
    }
  }

  trimWhiteSpaces = csvRows =>
    csvRows.map(row => row.map(value => value.trim()));

  mapCSVColumnNameToKey = headerRow => {
    headerRow[headerRow.findIndex(k => k === __('Name'))] = 'name';
    headerRow[headerRow.findIndex(k => k === __('Host'))] = 'host';
    headerRow[headerRow.findIndex(k => k === __('Provider'))] = 'provider';
    headerRow[headerRow.findIndex(k => k === __('UID'))] = 'uid_ems';
    return headerRow;
  };

  render() {
    return (
      <React.Fragment>
        <Dropzone
          accept=".csv"
          className="csv-upload-dropzone"
          activeClassName="active"
          onDrop={this.onFileDrop}
          disableClick
          disablePreview
          ref={node => {
            dropzoneRef = node; // See comment at top, this is so we can call dropzoneRef.open below
          }}
        >
          <EmptyState>
            <EmptyState.Icon type="pf" name="import" />
            <EmptyState.Title />
            <EmptyState.Info>
              {__('Import a file including a list of VMs to be migrated.')}
            </EmptyState.Info>
            <EmptyState.Action>
              <Button
                bsStyle="primary"
                onClick={() => {
                  dropzoneRef.open();
                }}
              >
                {__('Import')}
              </Button>
            </EmptyState.Action>
          </EmptyState>
        </Dropzone>
      </React.Fragment>
    );
  }
}

CSVDropzoneField.propTypes = {
  onCSVParseSuccess: PropTypes.func,
  onCSVParseFailure: PropTypes.func
};

export default CSVDropzoneField;
