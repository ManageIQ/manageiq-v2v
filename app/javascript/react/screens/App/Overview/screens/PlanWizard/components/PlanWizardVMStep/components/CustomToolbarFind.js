import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { bindMethods, noop, Button, Icon, FormControl } from 'patternfly-react';

/**
 * This is a custom toolbar find w/ limited functionality of Toolbar.Find
 * Note: does not implement the "findNext" and "findPrevious" functionality
 *
 * This should be added in a future release.
 */
class CustomToolbarFind extends React.Component {
  constructor() {
    super();
    this.state = {
      dropdownShown: false,
      currentValue: ''
    };

    bindMethods(this, [
      'toggleDropdownShown',
      'hideDropdown',
      'onValueKeyPress',
      'handleValueChange'
    ]);
  }

  onValueKeyPress(keyEvent) {
    const { onEnter } = this.props;
    const { currentValue } = this.state;

    if (keyEvent.key === 'Enter' && onEnter) {
      onEnter(currentValue);
    }
  }

  handleValueChange(event) {
    const { onChange } = this.props;

    this.setState({ currentValue: event.target.value });

    if (onChange) {
      onChange(event.target.value);
    }
  }

  hideDropdown() {
    const { onExit } = this.props;
    this.setState({ dropdownShown: false });
    onExit();
  }

  toggleDropdownShown() {
    this.setState(prevState => ({ dropdownShown: !prevState.dropdownShown }));
  }

  render() {
    const { dropdownShown, currentValue } = this.state;
    const { className, placeholder } = this.props;

    const classes = classNames('form-group toolbar-pf-find', className);

    const dropdownClasses = classNames('find-pf-dropdown-container', {
      show: dropdownShown
    });

    return (
      <div className={classes}>
        <Button
          bsStyle="link"
          className="btn-find"
          onClick={this.toggleDropdownShown}
        >
          <Icon type="fa" name="search" />
        </Button>
        <div className={dropdownClasses}>
          <FormControl
            type="text"
            id="find"
            value={currentValue}
            placeholder={placeholder}
            onKeyPress={e => this.onValueKeyPress(e)}
            onChange={this.handleValueChange}
          />
          <div className="find-pf-buttons">
            {/* {this.renderCounts()} */}
            <Button
              bsStyle="link"
              className="btn-find-close"
              onClick={this.hideDropdown}
            >
              <Icon type="pf" name="close" />
            </Button>
          </div>
        </div>
      </div>
    );
  }
}

CustomToolbarFind.propTypes = {
  /** Additional css classes */
  className: PropTypes.string,
  /** Placeholder text when empty */
  placeholder: PropTypes.string,
  /** Callback function when user hits the enter key */
  onEnter: PropTypes.func,
  /** Callback function when the find value changes */
  onChange: PropTypes.func,
  /** Callback function when exit clicked */
  onExit: PropTypes.func
};

CustomToolbarFind.defaultProps = {
  className: '',
  placeholder: '',
  onEnter: noop,
  onChange: noop,
  onExit: noop
};

export default CustomToolbarFind;
