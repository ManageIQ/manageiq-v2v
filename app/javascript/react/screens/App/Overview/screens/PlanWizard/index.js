import { connect } from 'react-redux';
import PlanWizard from './PlanWizard';
import * as PlanWizardActions from './PlanWizardActions';
import {
  planWizardOverviewFilter,
  planWizardFormFilter
} from './PlanWizardSelectors';
import { setMigrationsFilterAction } from '../../OverviewActions';

import reducer from './PlanWizardReducer';

export const reducers = { planWizard: reducer };

const mapStateToProps = ({ overview, planWizard, form }, ownProps) => {
  const selectedOverview = planWizardOverviewFilter(overview);
  const selectedForms = planWizardFormFilter(form);
  return {
    ...selectedOverview,
    ...selectedForms,
    ...ownProps.data
  };
};

const actions = {
  ...PlanWizardActions,
  setMigrationsFilterAction
};

const mergeProps = (stateProps, dispatchProps, ownProps) =>
  Object.assign(stateProps, ownProps.data, dispatchProps);

export default connect(mapStateToProps, actions, mergeProps)(PlanWizard);
