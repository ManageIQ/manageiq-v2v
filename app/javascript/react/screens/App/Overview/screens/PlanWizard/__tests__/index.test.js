import React from 'react';
import thunk from 'redux-thunk';
import promiseMiddleware from 'redux-promise-middleware';
import { createStore, applyMiddleware, combineReducers } from 'redux';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { initialState } from '../../../overview.fixtures';
import PlanWizard from '../PlanWizard';
import PlanWizardContainer from '../index';
import { reducers } from '../../../index';

import { coreComponents } from '../../../../../../../components/';
import componentRegistry from '../../../../../../../components/componentRegistry';

jest.mock('../../../../../../../components/componentRegistry');
componentRegistry.registerMultiple(coreComponents);

describe('Plan Wizard integration test', () => {
  const middlewares = [thunk, promiseMiddleware()];
  const generateStore = () =>
    createStore(
      combineReducers({ ...reducers }),
      {
        overview: initialState,
        form: {
          planWizardGeneralStep: {},
          planWizardCSVStep: {}
        }
      },
      applyMiddleware(...middlewares)
    );

  const mountComponent = store =>
    mount(
      <Provider store={store}>
        <PlanWizardContainer />
      </Provider>
    );

  it('should mount the PlanWizard with mapStateToProps reduced', () => {
    const store = generateStore();
    const wrapper = mountComponent(store);

    // query the unconnected component to assert reduced props
    const component = wrapper.find(PlanWizard);

    expect(component.props()).toMatchSnapshot();
  });
});
