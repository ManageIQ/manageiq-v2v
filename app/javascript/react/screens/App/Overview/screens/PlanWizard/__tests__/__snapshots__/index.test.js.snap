// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plan Wizard integration test should mount the PlanWizard with mapStateToProps reduced 1`] = `
Object {
  "hidePlanWizard": false,
  "hidePlanWizardAction": [Function],
  "planWizardExitedAction": [Function],
  "planWizardGeneralStep": Object {},
  "planWizardOptionsStep": Object {},
  "planWizardVMStep": Object {},
  "resetVmStepAction": [Function],
  "setMigrationsFilterAction": [Function],
  "setPlanScheduleAction": [Function],
  "setPlansBodyAction": [Function],
}
`;
